### PR TITLE
Support C++ (and C with GCC extension) global ctors/dtors

### PR DESCRIPTION
**Pull request by roscopeco**
This PR adds the most basic support for C++ user code (basically, removing the global constructor/destructor footgun by supporting the basics of the Itanium CXX ABI as used by GCC-m68k: https://itanium-cxx-abi.github.io/cxx-abi/abi.html#dso-dtor-runtime-api).

It also adds basic support for `__cxa_atexit` which G++ uses to register C++ destructors. This is not (currently) exposed to user code, but could quite easily be used to support user-code `atexit` handlers if desired.

There is a new test (in `code/software/tests/cpptest`) that tests the new functionality. Specifically, it exercises:

* C++ global constructors / destructors
* GCC `__attribute__((constructor))` and `__attribute__((destructor))` functions
  * With and without explicit priority

I have also run tests locally that check this doesn't break any existing code (that does not make use of these features) and am satisfied it doesn't.

Proof of life / Works On My Machine™:

Screenshot 2021-11-11 at 20 58 13

### CODE REVIEW

**Comment by 0xTJ** [comment]
Very cool work! It's nice to see it working! One little comment I have is that currently `exit()` just warm-boots, while *I think* it should call the destructors. Is that something you think should be added, or are good without for now?

File: code/software/libs/src/start_serial/init.S
Comment (0xTJ) [inline code/software/libs/src/start_serial/init.S:71]: A very minor thing, but I think this comment should read `__cxa_finalize`.

Comment (roscopeco) [reply code/software/libs/src/start_serial/init.S:71]: > Very cool work! It's nice to see it working! One little comment I have is that currently `exit()` just warm-boots, while _I think_ it should call the destructors. Is that something you think should be added, or are good without for now?

Ah yes, excellent point, good spot. I'll look into a way to make this happen cleanly 👍

Comment (roscopeco) [reply code/software/libs/src/start_serial/init.S:71]: > A very minor thing, but I think this comment should read `__cxa_finalize`.

Yup, it should. Will fix, thanks :)

Comment (roscopeco) [reply code/software/libs/src/start_serial/init.S]: Having looked into `exit` calling destructors, I've found that it's a bit tricky. For that reason, I'm proposing that we wait until `r2-initial-import` is merged before fixing (and accept that it will not work as planned with this merge - to be fair, it's already broken with HUGEROM, as noted at the bottom of the ticket linked below).

I've raised https://github.com/rosco-m68k/rosco_m68k/issues/269 to cover fixing this properly (post this merge and `r2-initial-import`) and will do the work as soon as the other branch is merged.

Let me know (or just resolve this conversation) if that's acceptable. Or if you can think of a better way to solve it, let me know :)

Comment (roscopeco) [reply code/software/libs/src/start_serial/init.S]: (I fixed the easy comment issue btw :sweat_smile: )

Comment (0xTJ) [reply code/software/libs/src/start_serial/init.S]: Makes sense to me!

**Review by 0xTJ — APPROVED** [review]
Looks good to me, replicated results as expected. :)

**Pull request merged**
